### ISSUE TITLE
Comments in empty function/loop/if-statement bodies not found

### ISSUE THREAD

**Issue by ec-m** [dave/dst#48]
Comments that are placed inside empty function, loop or if statement bodies are not found in the respective node's decorations. I would expect them to be placed somewhere in the e.g. `dst.FuncDeclDecorations` of the function or similarly the `dst.ForStmtDecorations` of the loop declaration. This snipped demonstrates the issue:

	package main

	func main() {
	}

	func empty() {
		// inside empty function, not found
	}

	func foo() {
		i := 1
		for i < 2 {
			// inside empty for statement, not found
		}
		// after empty for statement, found and associated with for statement

		if i == 3 {
			// inside empty if statement, not found
		}
		i = 4
	}

And here you can find code to reproduce the issue: https://play.golang.org/p/Yg9OOkoP6IF [**Disclaimer**: sometimes the playground needs multiple tries to import all packages, so if you encounter a `timeout running go build` error, just execute the code again.]

I tried using `dst` in order to fix the issue https://github.com/golang/go/issues/39753 I had with `go/ast`. As I am a fairly new user of this module, I might also have overlooked something - Please let me know if this is the case and how I could get the correct comment associations :) Thanks!

<pre>
$ go version
go version go1.14.2 darwin/amd64
</pre>

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/.../Library/Caches/go-build"
GOENV="/Users/.../Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/.../go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/.../go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/yb/hqncynqs0b5_3hyxcjpdsyr00000gr/T/go-build447564702=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

**Comment by dave** [comment]
I'll take a look at this now...

**Comment by dave** [comment]
Yeah if there's no decoration point after the comment, it'll attach it to the previous attachment point, which in the "empty function" example is `f1.Body.Decs.Lbrace`.

**Comment by dave** [comment]
Same with the for statement, in `forStmt.Body.Decs.Lbrace`.

**Comment by dave** [comment]
Take a look at [positions.go](https://github.com/dave/dst/blob/master/gendst/data/positions.go) in the gendst package... it shows exactly where each decoration point in on each node type. Might help track things down in future...



...

**Comment by dave** [comment]
e.g.:

```go
	// BlockStmt(0)
	if true /*Start*/ { /*Lbrace*/
		i++
	} /*End*/

	// BlockStmt(1)
	func() /*Start*/ { /*Lbrace*/ i++ } /*End*/ ()
```

**Comment by dave** [comment]
The algorithm that associates comments with decoration attachment points is one of those things that doesn't have a well defined correct answer... so sometimes you get situations like this where it's not exactly straightforward.

**Comment by ec-m** [comment]
Alright, thanks for the clarification! This helps a lot :) One last thing: Does `dst` provide a comment map like `ast.CommentMap` such that I can iterate over all comments that exists and see where they are attached to?

**Comment by dave** [comment]
Hmm ... just been looking at this and it's not easy in a generalised way. Let me take another look...

**Comment by dave** [comment]
Yeah you'd probably want something like this:

https://play.golang.org/p/NhLPbXeTCGo

... that `getDecorationInfo` function is copied from [here](https://github.com/dave/dst/blob/master/decorator/decorator-info-generated.go)... this is a valid use case that I would like to support... so I'm considering making it public... Not sure if it belongs in the `decorator` package or in the `dstutil` package. hmm

**Comment by ec-m** [comment]
Perfect! Thanks a lot for the help :)

**Comment by dave** [comment]
Hey Eva, I've just released v0.25.3 which makes that helper function public in the dstutil package. [See documentation here](https://github.com/dave/dst#dstutildecorations).

**Comment by ec-m** [comment]
Awesome! Thanks 👍
